mma: panic in sortTargetCandidateSetAndPick

This panic could trigger in production 

https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go#L506-L509.

`computeCandidatesForReplicaTransfer` returns an empty candidate set and bails out early if 
```
if sheddingSLS.sls <= loadNoChange && sheddingSLS.nls <= loadNoChange {
```

Thus, we could hit this panic when `sls <= loadNoChange` but `sheddingSLS > loadNoChange`. `sls.sls = loadNoChange` would be passed to `sortTargetCandidateSetAndPick` which then panics.  

```
panic: loadThreshold must be > loadNoChange

goroutine 1864 gp=0xc01672a780 m=5 mp=0xc00010d808 [running]:
panic({0x6a42e00?, 0x8934f50?})
	GOROOT/src/runtime/panic.go:802 +0x168 fp=0xc0801c2db0 sp=0xc0801c2d00 pc=0x4fa288
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/mmaprototype.sortTargetCandidateSetAndPick({0x89a8ab0, 0xc067c02140}, {{0xc065cdfea0, 0x0?, 0x4}, 0xc0833ce9b8?}, 0x1, 0x2, 0x0, 0xc010961980, ...)
	pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go:508 +0x1319 fp=0xc0801c3038 sp=0xc0801c2db0 pc=0x4603999
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/mmaprototype.(*rebalanceEnv).rebalanceReplicas(0xc067c02000, {0x89a8ab0, 0xc067c02140}, {0x2, {0x0, 0x3, 0x3, {0x3, 0x3, 0x1}, ...}}, ...)
	pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go:576 +0x136b fp=0xc0801c3730 sp=0xc0801c3038 pc=0x4613e0b
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/mmaprototype.(*rebalanceEnv).rebalanceStore(0xc067c02000, {0x89a8ab0, 0xc067c02140}, {0x2, {0x0, 0x3, 0x3, {0x3, 0x3, 0x1}, ...}}, ...)
	pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go:423 +0x6d3 fp=0xc0801c3928 sp=0xc0801c3730 pc=0x4612613
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/mmaprototype.(*rebalanceEnv).rebalanceStores(0xc067c02000, {0x89a8ab0, 0xc00b8bdd60}, 0x5)
	pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go:269 +0xecb fp=0xc0801c3c88 sp=0xc0801c3928 pc=0x4611bcb
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/mmaprototype.(*allocatorState).ComputeChanges(0xc00edf7ea0, {0x89a8ab0, 0xc00b8bdd60}, 0xc0ba805b00, {0x17f78f0?, 0xc0?, 0x0?})
	pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go:324 +0x2e6 fp=0xc0801c3d60 sp=0xc0801c3c88 pc=0x4601ee6
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*mmaStoreRebalancer).rebalance(0xc00b8f8090, {0x89a8ab0, 0xc00b8bdd60}, 0x1)
	pkg/kv/kvserver/mma_store_rebalancer.go:138 +0x183 fp=0xc0801c3e50 sp=0xc0801c3d60 pc=0x480df63
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*mmaStoreRebalancer).run(0xc00b8f8090, {0x89a8ab0, 0xc00b8bdd60}, 0xc004894360)
	pkg/kv/kvserver/mma_store_rebalancer.go:92 +0x32c fp=0xc0801c3f28 sp=0xc0801c3e50 pc=0x480db8c
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*mmaStoreRebalancer).start.func1({0x89a8ab0?, 0xc00b8bdd60?})
	pkg/kv/kvserver/mma_store_rebalancer.go:121 +0x2e fp=0xc0801c3f58 sp=0xc0801c3f28 pc=0x480ddae
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x89a8ab0, 0xc00b8bdd60})
	pkg/util/stop/stopper.go:440 +0x8a fp=0xc0801c3fc0 sp=0xc0801c3f58 pc=0x152824a
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.gowrap1()
	pkg/util/stop/stopper.go:441 +0x28 fp=0xc0801c3fe0 sp=0xc0801c3fc0 pc=0x1528188
runtime.goexit({})
	src/runtime/asm_amd64.s:1693 +0x1 fp=0xc0801c3fe8 sp=0xc0801c3fe0 pc=0x504e21
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 1542
	pkg/util/stop/stopper.go:438 +0x125

goroutine 1 gp=0xc0000061e0 m=nil [select, 17 minutes]:
runtime.gopark(...)
```

Epic CRDB-55052

--- a/pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go
+++ b/pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go
@@ -947,8 +947,10 @@ func (cs *clusterState) computeCandidatesForReplicaTransfer(
 	}
 
 	sheddingSLS = cs.computeLoadSummary(ctx, loadSheddingStore, &effectiveMeans.storeLoad, &effectiveMeans.nodeLoad)
-	if sheddingSLS.sls <= loadNoChange && sheddingSLS.nls <= loadNoChange {
-		// In this set of stores, this store no longer looks overloaded.
+	if sheddingSLS.sls <= loadNoChange {
+		// In this set of stores, this store no longer looks overloaded. Note that
+		// we don't consider nls here: if nls is high but sls is low, it means other
+		// stores on the node are causing node-level overload, not this store.
 		passObs.replicaShed(notOverloaded)
 		return candidateSet{}, sheddingSLS
 	}

--- a/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_test.go
+++ b/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_test.go
@@ -413,7 +413,14 @@ func TestClusterState(t *testing.T) {
 
 			// Recursively invoked in `include` directive.
 			var invokeFn func(t *testing.T, d *datadriven.TestData) string
-			invokeFn = func(t *testing.T, d *datadriven.TestData) string {
+			invokeFn = func(t *testing.T, d *datadriven.TestData) (output string) {
+				// Catch panics and return them as output instead of failing the test.
+				// This allows us to write regression tests for panics.
+				defer func() {
+					if r := recover(); r != nil {
+						output = fmt.Sprintf("panic: %v", r)
+					}
+				}()
 				// Start a recording span for each command. Commands that want to
 				// include the trace in their output can call finishAndGet().
 				ctx, finishAndGet := tracing.ContextWithRecordingSpan(

--- a/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_nls_sls_mismatch.txt
+++ b/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_nls_sls_mismatch.txt
@@ -1,0 +1,68 @@
+# Regression test for issue #160569: panic in sortTargetCandidateSetAndPick
+# when sls <= loadNoChange but nls > loadNoChange.
+#
+# Setup: Node n1 has two stores:
+# - s1: low CPU, moderate writes
+# - s2: very high CPU, replica disposition set to "refusing" (excluded from candidate set)
+# Node n2 has s3.
+#
+# When we try to shed from s1:
+# - Store mean over {s1, s3} (excluding s2 due to disposition)
+# - s1's store-level load relative to this mean is <= loadNoChange
+# - But n1's node CPU (s1+s2) is high, making nls > loadNoChange
+#
+# Before fix: computeCandidatesForReplicaTransfer doesn't bail out when
+# sls <= loadNoChange && nls > loadNoChange, then passes sls as loadThreshold
+# to sortTargetCandidateSetAndPick which requires loadThreshold > loadNoChange.
+
+set-store
+  store-id=1 node-id=1
+  store-id=2 node-id=1
+  store-id=3 node-id=2
+----
+node-id=1 locality-tiers=node=1
+  store-id=1 attrs=
+  store-id=2 attrs=
+node-id=2 locality-tiers=node=2
+  store-id=3 attrs=
+
+# s2 refuses replicas, so it will be filtered out of the candidate set
+# but its CPU still contributes to n1's node load.
+set-store-status store-id=2 replicas=refusing
+----
+ok refusing=replicas
+
+# 8-vCPU machines: capacity = 8e9 ns/s
+# s1: 1 vCPU (1B ns/s), 50% write bandwidth - at cluster level, WriteBandwidth is overloaded
+#     (because cluster mean includes s2's low writes, bringing mean down)
+# s2: 6 vCPU (6B ns/s), 10% writes - makes n1's total CPU = 7 vCPU
+# s3: 1 vCPU (1B ns/s), 50% write bandwidth
+#
+# Cluster store write mean = (50M + 10M + 50M) / 3 = 36.7M
+# s1's 50M is ~36% above mean => overloadSlow for WriteBandwidth
+#
+# Candidate set {s1, s3} store write mean = (50M + 50M) / 2 = 50M
+# s1's 50M equals mean => loadNoChange for WriteBandwidth
+#
+# Node CPU: n1 = 7B (s1+s2), n2 = 1B
+# Node mean = (7B + 1B) / 2 = 4B
+# n1's 7B is 75% above mean => overloadSlow for node CPU
+store-load-msg
+  store-id=1 node-id=1 load=[1000000000,50000000,0] capacity=[8000000000,100000000,1000000000] secondary-load=0 load-time=0s
+  store-id=2 node-id=1 load=[6000000000,10000000,0] capacity=[8000000000,100000000,1000000000] secondary-load=0 load-time=0s
+  store-id=3 node-id=2 load=[1000000000,50000000,0] capacity=[8000000000,100000000,1000000000] secondary-load=0 load-time=0s
+----
+
+# Single-replica range on s1.
+store-leaseholder-msg
+store-id=1
+  range-id=1 load=[100000000,10000000,0] raft-cpu=50000000
+    store-id=1 replica-id=1 type=VOTER_FULL leaseholder=true
+    config=num_replicas=1 constraints={} voter_constraints={}
+----
+
+# Before fix: panics in sortTargetCandidateSetAndPick
+# After fix: s1 should not try to shed since sls <= loadNoChange
+rebalance-stores store-id=1
+----
+panic: loadThreshold must be > loadNoChange

--- a/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_nls_sls_mismatch.txt
+++ b/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_nls_sls_mismatch.txt
@@ -65,4 +65,45 @@ store-id=1
 # After fix: s1 should not try to shed since sls <= loadNoChange
 rebalance-stores store-id=1
 ----
-panic: loadThreshold must be > loadNoChange
+[mmaid=1] rebalanceStores begins
+[mmaid=1] cluster means: (stores-load [cpu:2.7s/s, write-bandwidth:37 MB/s, byte-size:0 B]) (stores-capacity [cpu:8s/s, write-bandwidth:100 MB/s, byte-size:1.0 GB]) (nodes-cpu-load 4000000000) (nodes-cpu-capacity 12000000000)
+[mmaid=1] load summary for dim=CPURate (s1): loadLow, reason: load is >10% below mean [load=1000000000 meanLoad=2666666666 fractionUsed=12.50% meanUtil=33.33% capacity=8000000000]
+[mmaid=1] load summary for dim=WriteBandwidth (s1): overloadSlow, reason: fractionUsed < 75% [load=50000000 meanLoad=36666666 fractionUsed=50.00% meanUtil=36.67% capacity=100000000]
+[mmaid=1] load summary for dim=ByteSize (s1): loadNormal, reason: load is within 5% of mean [load=0 meanLoad=0 fractionUsed=0.00% meanUtil=0.00% capacity=1000000000]
+[mmaid=1] load summary for dim=CPURate (n1): overloadSlow, reason: fractionUsed < 75% [load=7000000000 meanLoad=4000000000 fractionUsed=43.75% meanUtil=33.33% capacity=16000000000]
+[mmaid=1] evaluating s1: node load overloadSlow, store load overloadSlow, worst dim WriteBandwidth
+[mmaid=1] overload-continued s1 ((store=overloadSlow worst=WriteBandwidth cpu=loadLow writes=overloadSlow bytes=loadNormal node=overloadSlow high_disk=false frac_pending=0.00,0.00(true))) - within grace period
+[mmaid=1] store s1 was added to shedding store list
+[mmaid=1] load summary for dim=CPURate (s2): overloadSlow, reason: fractionUsed < 75% and >1.75x meanUtil [load=6000000000 meanLoad=2666666666 fractionUsed=75.00% meanUtil=33.33% capacity=8000000000]
+[mmaid=1] load summary for dim=WriteBandwidth (s2): loadLow, reason: load is >10% below mean [load=10000000 meanLoad=36666666 fractionUsed=10.00% meanUtil=36.67% capacity=100000000]
+[mmaid=1] load summary for dim=ByteSize (s2): loadNormal, reason: load is within 5% of mean [load=0 meanLoad=0 fractionUsed=0.00% meanUtil=0.00% capacity=1000000000]
+[mmaid=1] load summary for dim=CPURate (n1): overloadSlow, reason: fractionUsed < 75% [load=7000000000 meanLoad=4000000000 fractionUsed=43.75% meanUtil=33.33% capacity=16000000000]
+[mmaid=1] evaluating s2: node load overloadSlow, store load overloadSlow, worst dim CPURate
+[mmaid=1] overload-continued s2 ((store=overloadSlow worst=CPURate cpu=overloadSlow writes=loadLow bytes=loadNormal node=overloadSlow high_disk=false frac_pending=0.00,0.00(true))) - within grace period
+[mmaid=1] store s2 was added to shedding store list
+[mmaid=1] load summary for dim=CPURate (s3): loadLow, reason: load is >10% below mean [load=1000000000 meanLoad=2666666666 fractionUsed=12.50% meanUtil=33.33% capacity=8000000000]
+[mmaid=1] load summary for dim=WriteBandwidth (s3): overloadSlow, reason: fractionUsed < 75% [load=50000000 meanLoad=36666666 fractionUsed=50.00% meanUtil=36.67% capacity=100000000]
+[mmaid=1] load summary for dim=ByteSize (s3): loadNormal, reason: load is within 5% of mean [load=0 meanLoad=0 fractionUsed=0.00% meanUtil=0.00% capacity=1000000000]
+[mmaid=1] load summary for dim=CPURate (n2): loadLow, reason: load is >10% below mean [load=1000000000 meanLoad=4000000000 fractionUsed=12.50% meanUtil=33.33% capacity=8000000000]
+[mmaid=1] evaluating s3: node load loadLow, store load overloadSlow, worst dim WriteBandwidth
+[mmaid=1] overload-continued s3 ((store=overloadSlow worst=WriteBandwidth cpu=loadLow writes=overloadSlow bytes=loadNormal node=loadLow high_disk=false frac_pending=0.00,0.00(true))) - within grace period
+[mmaid=1] store s3 was added to shedding store list
+[mmaid=1] start processing shedding store s1: cpu node load overloadSlow, store load overloadSlow, worst dim WriteBandwidth
+[mmaid=1] top-K[WriteBandwidth] ranges for s1 with lease on local s1: r1:[cpu:100ms/s, write-bandwidth:10 MB/s, byte-size:0 B]
+[mmaid=1] skipping lease shedding for calling store s1: not cpu overloaded: loadLow
+[mmaid=1] attempting to shed replicas next
+[mmaid=1] skipping s2 for replica transfer: replica disposition refusing (health ok)
+[mmaid=1] pre-means filtered 1 stores → remaining [1 3], means: store={[1000000000 50000000 0] [8000000000 100000000 1000000000] [0.125 0.5 0] [0 0]} node={4000000000 12000000000 0.3333333333333333}
+[mmaid=1] load summary for dim=CPURate (s1): loadNormal, reason: load is within 5% of mean [load=1000000000 meanLoad=1000000000 fractionUsed=12.50% meanUtil=12.50% capacity=8000000000]
+[mmaid=1] load summary for dim=WriteBandwidth (s1): loadNormal, reason: load is within 5% of mean [load=50000000 meanLoad=50000000 fractionUsed=50.00% meanUtil=50.00% capacity=100000000]
+[mmaid=1] load summary for dim=ByteSize (s1): loadNormal, reason: load is within 5% of mean [load=0 meanLoad=0 fractionUsed=0.00% meanUtil=0.00% capacity=1000000000]
+[mmaid=1] load summary for dim=CPURate (n1): overloadSlow, reason: fractionUsed < 75% [load=7000000000 meanLoad=4000000000 fractionUsed=43.75% meanUtil=33.33% capacity=16000000000]
+[mmaid=1] considering replica-transfer r1 from s1: store load [cpu:1s/s, write-bandwidth:50 MB/s, byte-size:0 B]
+[mmaid=1] candidates are:
+[mmaid=1] result(failed): no candidates found for r1 after exclusions
+[mmaid=1] start processing shedding store s2: cpu node load overloadSlow, store load overloadSlow, worst dim CPURate
+[mmaid=1] no top-K[CPURate] ranges found for s2 with lease on local s1
+[mmaid=1] start processing shedding store s3: cpu node load loadLow, store load overloadSlow, worst dim WriteBandwidth
+[mmaid=1] no top-K[WriteBandwidth] ranges found for s3 with lease on local s1
+[mmaid=1] rebalancing pass failures (store,reason:count): (s1,not-overloaded:1)
+pending(0)

